pkg/scalers/graphite_scaler.go

```diff
@@ -0,0 +1,206 @@
+		meta.from = val
+	} else {
+		return nil, fmt.Errorf("no %s given", grapqueryTime)
+	}
+
+	if val, ok := config.TriggerMetadata[grapThreshold]; ok && val != "" {
+		t, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing %s: %s", grapThreshold, err)
+		}
+
+		meta.threshold = t
+	}
+
+	authModes, ok := config.TriggerMetadata["authModes"]
+	// no authMode specified
+	if !ok {
+		return &meta, nil
+	}
+
+	authTypes := strings.Split(authModes, ",")
+	for _, t := range authTypes {
+		authType := authentication.Type(strings.TrimSpace(t))
+		switch authType {
+		case authentication.BasicAuthType:
+			if len(config.AuthParams["username"]) == 0 {
+				return nil, errors.New("no username given")
+			}
+
+			meta.username = config.AuthParams["username"]
+			// password is optional. For convenience, many application implement basic auth with
+			// username as apikey and password as empty
+			meta.password = config.AuthParams["password"]
+			meta.enableBasicAuth = true
+		default:
+			return nil, fmt.Errorf("err incorrect value for authMode is given: %s", t)
+		}
+	}
+
+	return &meta, nil
+}
+
+func (s *graphiteScaler) IsActive(ctx context.Context) (bool, error) {
+	val, err := s.ExecuteGrapQuery()
+	if err != nil {
+		graphiteLog.Error(err, "error executing graphite query")
+		return false, err
+	}
+
+	return val > 0, nil
+}
+
+func (s *graphiteScaler) Close() error {
+	return nil
+}
+
+func (s *graphiteScaler) GetMetricSpecForScaling() []v2beta2.MetricSpec {
+	targetMetricValue := resource.NewQuantity(int64(s.metadata.threshold), resource.DecimalSI)
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: kedautil.NormalizeString(fmt.Sprintf("%s-%s-%s", "graphite", s.metadata.serverAddress, s.metadata.metricName)),
+		},
+		Target: v2beta2.MetricTarget{
+			Type:         v2beta2.AverageValueMetricType,
+			AverageValue: targetMetricValue,
+		},
+	}
+	metricSpec := v2beta2.MetricSpec{
+		External: externalMetric, Type: externalMetricType,
+	}
+	return []v2beta2.MetricSpec{metricSpec}
+}
+
+func (s *graphiteScaler) ExecuteGrapQuery() (float64, error) {
+	client := &http.Client{}
+	queryEscaped := url_pkg.QueryEscape(s.metadata.query)
+	url := fmt.Sprintf("%s/render?from=%s&target=%s&format=json", s.metadata.serverAddress, s.metadata.from, queryEscaped)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return -1, err
+	}
+	if s.metadata.enableBasicAuth {
+		req.SetBasicAuth(s.metadata.username, s.metadata.password)
+	}
+	r, err := client.Do(req)
+	if err != nil {
+		return -1, err
+	}
+
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return -1, err
+	}
+	r.Body.Close()
+
+	result := gjson.GetBytes(b, "0.datapoints.#.0")
+	var v float64 = -1
+	for _, valur := range result.Array() {
+		if valur.String() != "" {
+			if float64(valur.Int()) > v {
+				v = float64(valur.Int())
+			}
+		}
+	}
```

Comment:
This appears to arbitrarily grab the first datapoint. I wonder if it might be better to return an error if the query returns more than one datapoint, similar to what the prometheus scaler does here: https://github.com/kedacore/keda/blob/main/pkg/scalers/prometheus_scaler.go#L254-L259.

Reply:
I'm actually not an expert in Graphite Querying it myself. Is it actually a thing to return a single datapoint? I've spent the last 45 minutes trying to teach myself graphite's query language, but I always seem to be returned many elements within the datapoint array.

Even looking at some established graphs for things like singleStat in grafana, I can see that queries return many elements within the datapoints array.

What we can do however is ensure we only are returned a single series

Reply:
Okay. Provided that the original bit about finding the largest datapoint was correct. This should now throw an error prometheus style if you are returned multiple series, rather than silently taking the first.
`			if float64(valur.Int()) > v {
				v = float64(valur.Int())
			}
`

  Warning  KEDAScalerFailed          4s (x2 over 9s)  keda-operator  graphite query https_metric.* returned multiple series

Reply:
I am seeing a bug though. But I'm not clear on where its coming from. I can see according to the hpa controller we're getting a very large number reported back as our metric.

  `
"graphite-http---graphite-default-svc-8089-https_metric" (target average value):  326192464 / 100
  `

The data being generated is with the following command to produce data between 1 and 10.
`
while true; do echo "https_metric $(shuf -i1-10 -n1) $(date +%s)" | nc -q 1 localhost 2003; echo "data sent :)"; sleep 1; done
`

As I step through the debugger on this, the graphite scaler is correctly only returning a float between 1 and 10, whatever the largest within its time window is. I don't see where that very large number is coming from. I stepped back to the original committed code prior to any of my own changes, and I appear to get the same behaviour

this is with scaledObject definition
`
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: graphite-scaledobject
spec:
  scaleTargetRef:
    name: php-apache-graphite
  minReplicaCount: 0
  maxReplicaCount: 5
  pollingInterval: 5
  cooldownPeriod:  10
  triggers:
  - type: graphite
    metadata:
      serverAddress: http://graphite.default.svc:8089
      metricName: https_metric
      threshold: '100'
      query: "https_metric"
      queryTime: '-30Seconds'

Reply:
After more investigation/messing around. I'm only confusing myself further

```
 curl localhost:8001/apis/external.metrics.k8s.io/v1beta1/namespaces/default/graphite-http---graphite-default-svc-8089-https_metric
{
  "kind": "ExternalMetricValueList",
  "apiVersion": "external.metrics.k8s.io/v1beta1",
  "metadata": {

  },
  "items": [
    {
      "metricName": "graphite-http---graphite-default-svc-8089-https_metric",
      "metricLabels": null,
      "timestamp": "2021-09-07T04:33:39Z",
      "value": "1630989210"
    }
  ]
}```

If I query the external metrics endpoint directly, the value is reported as the current timestamp.

I also can't actually seem to find the GetMetrics method ever being called. All invocations of executeGrapQuery are coming from the IsActive loop. I've tried my best to track through the logic of greater keda to figure out how we could possibly be returning the timestamp as a value, but I'm a bit lost at this point.

Reply:
> Provided that the original bit about finding the largest datapoint was correct.

I don't know if that was correct. The decision to find the largest datapoint seems arbitrary. What if you are trying to scale base on a the lower bound of a metric, like when the available quantity of a given resource falls _below_ a given threshold? Then returning the largest value would be incorrect. That is way I think it is better to require the provided query only return one or zero data points. This should be possible using an aggregate function in your Graphite query.

Reply:
The logic is fixed to just not return the latest datapoint instead of arbitrarily returning the largest.

I'm still not much closer however on debugging where that strange target average value `  "graphite-http---graphite-default-svc-8089-https_metric" (target average value):  326215938 / 100`
 
I know something is wrong, because even with my scaler code not running, the external metrics api is returning a value which should be impossible. The GetMetrics function is just never being called, yet we're getting data somehow, which I have no explanation for. I spent a couple hours digging through how all of the different pieces fit together and my best conclusion is that this issue is originating on the metrics server side. 

I'll need to figure out how I can run my debugger against that to really get to the bottom of this, I think. There are no instructions for that in keda, nor https://github.com/kubernetes-sigs/custom-metrics-apiserver/tree/2c1ad756d33c6596d8bd1cec9d2f7425b350f56a that I've been able to find, so that's something else to figure out first.

Reply:
I finally fixed a gross bug that has haunted me for the last three days. It was half the metrics apiserver, half me. https://github.com/kubernetes-sigs/custom-metrics-apiserver
It's one of those things that I finally only figured out while laying in bed at 2 am last night. There was no bug after all.
The keda project is two things, the operator and the custom metrics apiserver.
I didn't realize the scaler code is actually embedded in both the operator, and the apiserver component. Two different code paths are taken to execute my scaler code.

The keda project only contains instructions for really running the operator locally and that's what I had been debugging against which is why nothing was making any sense. I was testing one particular code path while a separate binary held the other, broken code path.
Two things confused me.

1) The lack of instructions for running the custom metrics server locally lead me to mostly discount it as a source of things to debug against.
2) Somehow thinking that the metrics apiserver was calling the GetMetrics() function from the operator code ( the operator does actually execute the scaler code but only for checking whether or not it should scale up from zero, not actually feeding data into the hpa loop)

My metrics api server was built and deployed only once using an old version of my code that did actually have a bug, and I just left it not knowing there was an actual significance in regularly going through the full multi minute dance of building a new container, publishing it, and deploying it to k8s. When iteration on the operator itself is very quick and local.

Reply:
Wow... Glad you got to the bottom of this. Using the scaler code in two separate binaries is definitely not something I would have expected.